sqs/src/main/scala/akka/stream/alpakka/sqs/SqsAckFlowStage.scala

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.sqs
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.stream.stage._
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+import com.amazonaws.{AmazonWebServiceResult, ResponseMetadata}
+import com.amazonaws.handlers.AsyncHandler
+import com.amazonaws.services.sqs.AmazonSQSAsync
+import com.amazonaws.services.sqs.model.{
+  DeleteMessageRequest,
+  DeleteMessageResult,
+  SendMessageRequest,
+  SendMessageResult
+}
+
+import scala.concurrent.{Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+class SqsAckFlowStage(queueUrl: String, sqsClient: AmazonSQSAsync)
+    extends GraphStage[FlowShape[MessageActionPair, Future[AmazonWebServiceResult[ResponseMetadata]]]] {
+
+  private val in = Inlet[MessageActionPair]("messages")
+  private val out = Outlet[Future[AmazonWebServiceResult[ResponseMetadata]]]("result")
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
+
+    val logic = new GraphStageLogic(shape) with StageLogging {
+      private var inFlight = new AtomicInteger(0)
+      @volatile private var inIsClosed = false
```

Comment:
Put all interaction in methods called from async callbacks instead of these thread primitives and make use of the guarantees streams gives you.